### ISSUE TITLE
Problem in shapefile saving when copying features from same layer

### ISSUE THREAD

**Issue by nprigour** [locationtech/udig-platform#260]
Hi,
 It seems there is a serious problem when someone attempts to copy from a shapefile layer a feature and paste it to the same layer. A subsequent commit never completes and seems to blocks indefinitely the UI.
This issue was not present in old udig 1.4 (geotools 9.0-M0 and seems to have occured after shifting to locationtech and newer version of geotools (geotools 14.1). 
How to reproduce:
1. select a feature on a shapefile layer and do a right click and select copy from the popup menu
2. click on the map (be sure same layer is selected),  unselect the feature and do again a right click and select paste from the popup menu
3. click the map commit button and observe that the commit progress monitor that appears never completes    
comment: it often occurs that another completely different feature is copied (based on selection and zoom o map after execution of step 2

**Comment by nprigour** [comment]
Hi @fgdrf, 
From a little further investigation it seems that the problem has to do with some read locks being held by another thread from the one that tries to commit (write) to shapefile. More specifically in org.geotools.data.shapefile.files.ShpFiles class (located in gt-shapefile.jar) when method acquireWrite(...) (starting at line 479) is called during commit when a write lock is attempted (see arrow in code provided below) the lock attempt blocks indefinitely.  

```
    /**
     * Acquire a URL for read and write purposes.
     * <p>
     * It is recommended that get*Stream or get*Channel methods are used when reading or writing to
     * the file is desired.
     * </p>
     * 
     * @see #getInputStream(ShpFileType, FileReader)
     * @see #getReadChannel(ShpFileType, FileReader)
     * @see #getWriteChannel(ShpFileType, FileReader)
     * 
     * 
     * @param type the type of the file desired.
     * @param requestor the object that is requesting the URL. The same object must release the lock
     *        and is also used for debugging.
     * @return the URL to the file of the type requested
     */
    public URL acquireWrite(ShpFileType type, FileWriter requestor) {
        URL url = urls.get(type);
        if (url == null) {
            return null;
        }

        // we need to give up all read locks before getting the write one
        Collection<ShpFilesLocker> threadLockers = getCurrentThreadLockers();
        relinquishReadLocks(threadLockers);
    -->    readWriteLock.writeLock().lock();
        threadLockers.add(new ShpFilesLocker(url, requestor));
        mapCache.cleanFileCache(url);
        return url;
    }
```
Checking with debug enabled it seems that there are 3 read lock still active from another thread in the shapefile (see screenshot below taking from elipse debug perspective) and I wonder why?
![default](https://user-images.githubusercontent.com/26147785/34581992-1bdc490c-f19b-11e7-9050-0c3cdeabb514.png)
The following image shows that the read lockers refer to .fix, .shp, .dbf files
![default](https://user-images.githubusercontent.com/26147785/34582236-e9e3fab6-f19b-11e7-8d03-69219c2413df.png)

From the above it seems that the issue may be related to geotools (is it a bug??) and the shapefile implementation used. Since I was working previously with udig 1.4 I never experienced such an issue. Of course there geotools 9.0-M0 were used and an IndexedShapefileDataStore which was removed later on (in present geotools 14.1 are used by udig and IndexedShapefileDataStore has been removed in favor of ShapefileDataStore). 
Of course all the above are just my observations and speculations. Do you know if anybody from the geotools project can be put into loop to investigate further because I got lost while trying to further debug it.

**Comment by fgdrf** [comment]
Let us ask @aaime and @jodygarnett who are familiar with GeoTools codebase and maybe can remember a bug like this in GeoTools Version 14.1. Do you have any ideas?

**Comment by jodygarnett** [comment]
We did a QA run on geotools datastores not closing file readers, think it was prior to geotools 14.1.

**Comment by nprigour** [comment]
The QA reveal that everything was handled correctly?
Note that: 
1. The problem occured when I tested with udig copy feature command and 
2. only in the case the source and destination shapefile layer for the feature was the same. 
3. When copying between 2 different shapefile layers or creating a new feature or deleting a feature via the map tools everything work correctly. Read locks were held and released correctly and write locks always were acquired.

**Comment by nprigour** [comment]
Problem identified see pull request #261

**Comment by fgdrf** [comment]
fixed while #261 was merged
